### ISSUE TITLE
Update API doesn't work anymore when index is not created yet

### ISSUE THREAD

**Issue by mouadino** [elastic/elasticsearch#19857]
**Elasticsearch version**: 2.3.3 and 2.1.1

**Plugins installed**: []

**JVM version**: 1.8.0_92

**OS version**: OS X El Capitan 10.11.5

**Description of the problem including expected versus actual behavior**:

When migrating an Elasticsearch cluster from 2.1.1 to 2.3.3, a previously working code started to failing and I looked to https://www.elastic.co/guide/en/elasticsearch/reference/current/breaking-changes-2.3.html trying to make sense of the failure sadly didn't see any mention of the change so I am unsure whether this is a regression bug or documentation issue.

The problem that I am referring to is when using the update API, which we use to do upsert of documents which works okay in  2.1.1 when index is not created already (given that `action.auto_create_index: true`) but fail with 2.3.3 with same configuration complaining that index was not found.

**Steps to reproduce**:
 
1. Using 2.3.3 run `curl -XPOST http://localhost:9200/index/doc_type/1/_update -d '{"doc": {"name" : "test"}, "doc_as_upsert": true}'`
2. Using 2.1.1 try same query as before.

With 2.3.3 you get:

```
{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index","resource.type":"index_expression","resource.id":"...","index":"..."}],"type":"index_not_found_exception","reason":"no such index","resource.type":"index_expression","resource.id":"...","index":"..."},"status":404}
```

With 2.1.1 it works as expected.

**Provide logs (if relevant)**:

```
[dummy-index] IndexNotFoundException[no such index]
	at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndices(IndexNameExpressionResolver.java:151)
	at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteIndices(IndexNameExpressionResolver.java:95)
	at org.elasticsearch.cluster.metadata.IndexNameExpressionResolver.concreteSingleIndex(IndexNameExpressionResolver.java:208)
	at org.elasticsearch.action.support.single.instance.TransportInstanceSingleOperationAction$AsyncSingleAction.doStart(TransportInstanceSingleOperationAction.java:138)
	at org.elasticsearch.action.support.single.instance.TransportInstanceSingleOperationAction$AsyncSingleAction.start(TransportInstanceSingleOperationAction.java:123)
	at org.elasticsearch.action.support.single.instance.TransportInstanceSingleOperationAction.doExecute(TransportInstanceSingleOperationAction.java:73)
	at org.elasticsearch.action.update.TransportUpdateAction.innerExecute(TransportUpdateAction.java:147)
	at org.elasticsearch.action.update.TransportUpdateAction.doExecute(TransportUpdateAction.java:142)
	at org.elasticsearch.action.update.TransportUpdateAction.doExecute(TransportUpdateAction.java:66)
	at org.elasticsearch.action.support.TransportAction.doExecute(TransportAction.java:149)
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:137)
	at org.elasticsearch.action.support.TransportAction.execute(TransportAction.java:85)
	at org.elasticsearch.client.node.NodeClient.doExecute(NodeClient.java:58)
	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:359)
	at org.elasticsearch.client.FilterClient.doExecute(FilterClient.java:52)
	at org.elasticsearch.rest.BaseRestHandler$HeadersAndContextCopyClient.doExecute(BaseRestHandler.java:83)
	at org.elasticsearch.client.support.AbstractClient.execute(AbstractClient.java:359)
	at org.elasticsearch.client.support.AbstractClient.update(AbstractClient.java:396)
	at org.elasticsearch.rest.action.update.RestUpdateAction.handleRequest(RestUpdateAction.java:126)
	at org.elasticsearch.rest.BaseRestHandler.handleRequest(BaseRestHandler.java:54)
	at org.elasticsearch.rest.RestController.executeHandler(RestController.java:205)
	at org.elasticsearch.rest.RestController.dispatchRequest(RestController.java:166)
	at org.elasticsearch.http.HttpServer.internalDispatchRequest(HttpServer.java:128)
	at org.elasticsearch.http.HttpServer$Dispatcher.dispatchRequest(HttpServer.java:86)
	at org.elasticsearch.http.netty.NettyHttpServerTransport.dispatchRequest(NettyHttpServerTransport.java:449)
	at org.elasticsearch.http.netty.HttpRequestHandler.messageReceived(HttpRequestHandler.java:61)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.elasticsearch.http.netty.pipelining.HttpPipeliningHandler.messageReceived(HttpPipeliningHandler.java:60)
	at org.jboss.netty.channel.SimpleChannelHandler.handleUpstream(SimpleChannelHandler.java:88)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.http.HttpChunkAggregator.messageReceived(HttpChunkAggregator.java:194)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.handler.codec.http.HttpContentDecoder.messageReceived(HttpContentDecoder.java:135)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:296)
	at org.jboss.netty.handler.codec.frame.FrameDecoder.unfoldAndFireMessageReceived(FrameDecoder.java:459)
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.callDecode(ReplayingDecoder.java:536)
	at org.jboss.netty.handler.codec.replay.ReplayingDecoder.messageReceived(ReplayingDecoder.java:435)
	at org.jboss.netty.channel.SimpleChannelUpstreamHandler.handleUpstream(SimpleChannelUpstreamHandler.java:70)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline$DefaultChannelHandlerContext.sendUpstream(DefaultChannelPipeline.java:791)
	at org.elasticsearch.common.netty.OpenChannelsHandler.handleUpstream(OpenChannelsHandler.java:75)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:564)
	at org.jboss.netty.channel.DefaultChannelPipeline.sendUpstream(DefaultChannelPipeline.java:559)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:268)
	at org.jboss.netty.channel.Channels.fireMessageReceived(Channels.java:255)
	at org.jboss.netty.channel.socket.nio.NioWorker.read(NioWorker.java:88)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.process(AbstractNioWorker.java:108)
	at org.jboss.netty.channel.socket.nio.AbstractNioSelector.run(AbstractNioSelector.java:337)
	at org.jboss.netty.channel.socket.nio.AbstractNioWorker.run(AbstractNioWorker.java:89)
	at org.jboss.netty.channel.socket.nio.NioWorker.run(NioWorker.java:178)
	at org.jboss.netty.util.ThreadRenamingRunnable.run(ThreadRenamingRunnable.java:108)
	at org.jboss.netty.util.internal.DeadLockProofWorker$1.run(DeadLockProofWorker.java:42)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
```

**Comment by jimferenczi** [comment]
I tested your recreation with ES 2.3.3 and it worked fine. The index is created as expected and it fails if `action.auto_create_index` is set to false.  I am closing the issue since I am not able to reproduce, feel free to reopen if you find a recreation that does not work.

**Comment by mouadino** [comment]
@jimferenczi Sadly I cannot reopen the issue but I am updating here with reproducible steps.

### Configuration

Install both elasticsearch versions 2.1.0 and 2.3.0, for each do the following:

1. Change configuration so that `action.auto_create_index: true` and `index.mapper.dynamic: false`. 
2. Trying to create now a document fail as expected, `curl -XPOST http://192.168.100.100:9200/twitter:0/tweet/1/_update -d '{"doc": {"message" : "test"}, "doc_as_upsert": true}'
{"error":{"root_cause":[{"type":"type_missing_exception","reason":"type[[tweet, trying to auto create mapping, but dynamic mapping is disabled]] missing","index":"twitter"}],"type":"type_missing_exception","reason":"type[[tweet, trying to auto create mapping, but dynamic mapping is disabled]] missing","index":"twitter"},"status":404}`
3. Create a template: `curl -XPUT --data "@mapping.json" http://192.168.100.100:9200/_template/twitter`

`mapping.json` contains:

```
{
  "template": "*",
  "order":0,
  "mappings": {
    "tweet": {
      "_source":{
        "enabled":true
      },
      "properties": {
        "message": {
          "type": "string"
        }
      }
    }
  }
}
```

### Issue:

Now this is what happen when doing an upsert with both elasticsearch versions:

**Elasticsearch 2.1.0:**

`curl -XPOST http://192.168.100.100:9200/twitter:0/tweet/1/_update -d '{"doc": {"message" : "test"}, "doc_as_upsert": true}'
{"_index":"twitter:0","_type":"tweet","_id":"1","_version":1,"_shards":{"total":2,"successful":1,"failed":0}}`

**Elasticsearch 2.3:**

`curl -XPOST http://192.168.100.100:9200/twitter:0/tweet/1/_update -d '{"doc": {"message" : "test"}, "doc_as_upsert": true}'
{"error":{"root_cause":[{"type":"index_not_found_exception","reason":"no such index","resource.type":"index_expression","resource.id":"twitter:0","index":"twitter:0"}],"type":"index_not_found_exception","reason":"no such index","resource.type":"index_expression","resource.id":"twitter:0","index":"twitter:0"},"status":404}`

Thanks again for looking at this bug

**Comment by clintongormley** [comment]
thanks for the recreation @mouadino 

I hoped this would be fixed by https://github.com/elastic/elasticsearch/pull/19478 but apparently not.  This still fails on 2.4.0.  

I think that theoretically this should work, but it might be complicated.  The logic should be:

* does the index exist?
* if not, can i create the index? (check auto create)
* is there a matching template?
* create the index with the mappings
* does the mapping exist?
* if not, can i create the mapping? (check index.mapper.dynamic)
* create the mapping

The downside here is that (with auto create true, and mapper dynamic false) you can end up creating an index even though you can't create the type.  But i think that makes sense.

@jimferenczi could you take a look please?

**Comment by jimferenczi** [comment]
> I hoped this would be fixed by #19478 but apparently not. This still fails on 2.4.0.

When set at the node level we don't check if a template would have matched.  https://github.com/elastic/elasticsearch/pull/19478 fixed the case where it is set at the index level for instance within the template:

````
{
  "template": "*",
  "order":0,
  "settings": {
    "index.mapper.dynamic": false
  },
  "mappings": {
    "tweet": {
      "_source":{
        "enabled":true
      },
      "properties": {
        "message": {
          "type": "string"
        }
      }
    }
  }
}
`````
... this works with the downside that @clintongormley mentioned:
> The downside here is that (with auto create true, and mapper dynamic false) you can end up creating an index even though you can't create the type.

Since it is not possible to set this setting at the node level in 5.x I don't think we should try to fix it in 2.x but rather document the alternative ?

**Comment by clintongormley** [comment]
> Since it is not possible to set this setting at the node level in 5.x I don't think we should try to fix it in 2.x but rather document the alternative ?

Ahhh thanks, I'd forgotten about that. Same reason i closed a similar issue previously.

+1 to docs
